### PR TITLE
GPT2 SDPA patterns mapping to ATen SDPA kernels

### PR DESCRIPTION
**Pull request by sanchitintel**
<strike>#121555 should be merged first</strike>

Graph breaks are introduced if the `inv_scale` is read with `.item()`, so not reusing its value but looked it up directly from https://github.com/huggingface/transformers/blob/main/src/transformers/models/gpt2/modeling_gpt2.py#L182-L220

It seems GPT2 can't directly use `is_causal` as `True` in `torch.nn.functional.scaled_dot_product_attention` for most cases.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Comment by sanchitintel** [comment]
@Valentine233, pattern 20 (both causal mask & attention mask being present) seems to need more investigation.
May we add it in a separate PR?
I combined patterns 18 & 19 into one, and opened #121866 for it instead. I'll seek your review on that PR after CI would complete.

Thanks!

**Comment by Valentine233** [comment]
> @Valentine233, pattern 20 (both causal mask & attention mask being present) seems to need more investigation. May we add it in a separate PR? I combined patterns 18 & 19 into one, and opened #121866 for it instead. I'll seek your review on that PR after CI would complete.
> 
> Thanks!

"pattern 20 (both causal mask & attention mask being present) seems to need more investigation." Do you indicate the UT problem?

**Comment by sanchitintel** [comment]
Sorry for my late response, @Valentine233!
It's the same issue that you had encountered locally with it (failing UT).
